### PR TITLE
samples: net: zperf: Support existing IP

### PR DESCRIPTION
**Pull request by bublover**
Considering that IP can be set dynamically by DHCP
or statically before test previously, we could have
alternative way by supporting both setting IP and
existing IP.

Signed-off-by: Bub Wei <bub.wei@unisoc.com>

### CODE REVIEW

**Review by pfalcon — DISMISSED** [review]
@bublover, Can you describe what problem this patch fixes? It goes pretty deep into netif internal structure, that should be avoided.

**Comment by bublover** [comment]
@pfalcon Hello, I didn't modify any netif internal structure.
I am working on UNISOC, and we are doing upstream for our chip uwp5661 which is Wi-Fi and BT chip in breif.
Now I am doing zpert test for Wi-Fi network. I don't know which IP can be set by DHCP server.  Current zperf has static setting IP only.  I am adding an alternative way and using exsiting IP is available in this patch.

**Comment by bublover** [comment]
> > I have uploaded patches to make zperf available.
> 
> Sure, great work, thanks! But each change is worth separate consideration, and sorry but I'm still not sure what's the purpose of this change.
> 
> > I didn't modify any netif internal structure.
> 
> You don't modify, but you access it with code like https://github.com/zephyrproject-rtos/zephyr/pull/11860/files#diff-f40b40ea715a0c81f30cff262ede9ef8R230 . Applications should avoid doing that, unless there's really good reason. And again, the reason for doing that so far not clear.
> 
> > I don't know which IP can be set by DHCP server. Current zperf has static setting IP only. I am adding an alternative way and using exsiting IP is available in this patch.
> 
> So, samples/net/zperf is just one of dozens of Zephyr's networking samples. All networking settings of any sample are configured using options available in https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/net/lib/config/Kconfig . Please see if it already contains setting which you need. If not, please explain what exactly you're trying to achieve, we'll consider if such an option needs to be added on `net/lib/config` level, or it's something zperf-specific.

Thanks for your comments.
What you mean is that the sample is supporting setting IP in Kconfig macro only, right?   But I don't know what IP is at compile time before wifi sta connects to an wifi AP.

**Comment by pfalcon** [comment]
Well, I see. This particular app has pretty adhoc (perhaps, very old) configuration, where it tries to set up its own IP address, etc. And now you need to work that around. But that doesn't seem good. Instead, we should clean up the app itself.

I'd like to hear @jukkar's comments here before proceeding further.

**Comment by bublover** [comment]
> Well, I see. This particular app has pretty adhoc (perhaps, very old) configuration, where it tries to set up its own IP address, etc. And now you need to work that around. But that doesn't seem good. Instead, we should clean up the app itself.
> 
> I'd like to hear @jukkar's comments here before proceeding further.

Well I'm looking forward to new iperf app. Upon this, I can be doing network test.

File: samples/net/zperf/src/zperf_tcp_receiver.c
Comment (jukkar) [inline samples/net/zperf/src/zperf_tcp_receiver.c]: No need to call strlen() here, just do `MY_IP4_ADDR[0]`

File: samples/net/zperf/src/zperf_tcp_receiver.c
Comment (jukkar) [inline samples/net/zperf/src/zperf_tcp_receiver.c]: same here

File: samples/net/zperf/src/zperf_tcp_receiver.c
Comment (jukkar) [inline samples/net/zperf/src/zperf_tcp_receiver.c]: indendation seems to be slighly off here. Please also check other shell_fprintf() commands as the indentation is off there too.

File: samples/net/zperf/src/zperf_tcp_receiver.c
Comment (jukkar) [inline samples/net/zperf/src/zperf_tcp_receiver.c]: indendation

File: samples/net/zperf/src/zperf_udp_receiver.c
Comment (jukkar) [inline samples/net/zperf/src/zperf_udp_receiver.c]: replace the strlen() here too

File: samples/net/zperf/src/zperf_udp_receiver.c
Comment (jukkar) [inline samples/net/zperf/src/zperf_udp_receiver.c]: same here

**Review by jukkar — CHANGES_REQUESTED** [review]

**Comment by jukkar** [comment]
> > Well, I see. This particular app has pretty adhoc (perhaps, very old) configuration, where it tries to set up its own IP address, etc. And now you need to work that around. But that doesn't seem good. Instead, we should clean up the app itself.
> > I'd like to hear @jukkar's comments here before proceeding further.
> 
> Well I'm looking forward to new iperf app. Upon this, I can be doing network test.

I am not aware of any new iperf app. The changes proposed here look just fine.

**Comment by pfalcon** [comment]
> The changes proposed here look just fine.

Well, they doesn't look fine. They look like many other things in Zephyr: there's technical debt, so people have to work it around, because resolving it hard. And thus the amount of technical debt only increases. But if the networking maintainer thinks that's "fine", I'm removing my -1, it's a sample after all, it's great that someone uses it and does something about it at all.

File: samples/net/zperf/src/zperf_tcp_receiver.c
Comment (bublover) [reply samples/net/zperf/src/zperf_tcp_receiver.c]: Thanks Jukkar,
With MY_IP4ADDR[0], compiled error is happening.
zephyr/samples/net/zperf/src/zperf_udp_receiver.c:329:30: warning: pointer of type ‘void *’ used in arithmetic [-Wpointer-arith]
   if (MY_IP4ADDR && MY_IP4ADDR[0]) {
                              ^
zephyr/samples/net/zperf/src/zperf_udp_receiver.c:329:30: error: void value not ignored as it ought to be
   if (MY_IP4ADDR && MY_IP4ADDR[0]) {

**Comment by jukkar** [comment]
> Well, they doesn't look fine

Huh, well if you want to rewrite the zperf sample app to use socket or something else, then just send PR instead of complaining here.

**Comment by bublover** [comment]
> > > Well, I see. This particular app has pretty adhoc (perhaps, very old) configuration, where it tries to set up its own IP address, etc. And now you need to work that around. But that doesn't seem good. Instead, we should clean up the app itself.
> > > I'd like to hear @jukkar's comments here before proceeding further.
> > 
> > 
> > Well I'm looking forward to new iperf app. Upon this, I can be doing network test.
> 
> I am not aware of any new iperf app. The changes proposed here look just fine.

About getting default iface IP,  we'd like to use user API instead of kernel struct. Do you have any plan or idea?

**Comment by pfalcon** [comment]
> Huh, well if you want to rewrite the zperf sample app to use socket or something else, then just send PR instead of complaining here.

There's no need to rewrite it using sockets. But yes, this old, unmaintained sample app exactly needs a good rewrite by now. And surely, I don't want to do that, just like everyone else. But pretending the problem doesn't exist won't do. We'll need to pay that technical debt eventually. Actually, we're paying interest every day already.

**Comment by jukkar** [comment]
> About getting default iface IP, we'd like to use user API instead of kernel struct. Do you have any plan or idea?

As I mentioned in the other PR that you sent yesterday, there is the net_if_ipv6_select_src_iface() and net_if_ipv4_select_src_iface() functions you might be able to use here.

**Comment by bublover** [comment]
> > About getting default iface IP, we'd like to use user API instead of kernel struct. Do you have any plan or idea?
> 
> As I mentioned in the other PR that you sent yesterday, there is the net_if_ipv6_select_src_iface() and net_if_ipv4_select_src_iface() functions you might be able to use here.

Thanks. I went through these APIs, they need dst in_addr. There is a problem. As zperf server, dst ip is unknown. And I tried to take gw ip as dst ip, but no API available (get gw ip by iface), only a set gw API.
Anyway, I will be looking into it. 

Thanks @pfalcon and you for the technical discussions and professional comments.

**Comment by pfalcon** [comment]
> Thanks @pfalcon and you for the technical discussions and professional comments.

I'm actually sorry that arguments between networking developers happen in discussion of your PR, but well, we need to discuss problem points. Please don't feel discouraged and continue work on this tool, it's better to do something to improve it rather than wait.

**Comment by jukkar** [comment]
> Thanks. I went through these APIs, they need dst in_addr. There is a problem. As zperf server, dst ip is unknown. And I tried to take gw ip as dst ip, but no API available (get gw ip by iface), only a set gw API.
> Anyway, I will be looking into it.

If the dst is not known, then you could use `net_ipv6_unspecified_address()` when calling the func and still get some meaningful results out.

**Review by bublover — COMMENTED** [review]
@jukkar Could you please review again?

File: samples/net/zperf/src/zperf_tcp_receiver.c
Comment (bublover) [reply samples/net/zperf/src/zperf_tcp_receiver.c]: If MY_IP4ADDR is NULL,  it's not allowed to access it with MY_IP4ADDR[0].

File: samples/net/zperf/src/zperf_shell.c
Comment (jukkar) [inline samples/net/zperf/src/zperf_shell.c]: The iface can be set to NULL in which case the function will search all the interfaces.

So your call could look like this
```
in4_addr = net_if_ipv4_select_src_addr(NULL, net_ipv4_unspecified_address());
```

I just noticed that const is missing from net_if_ipv4_select_src_addr() dst address, I will send a patch fixing it.

File: samples/net/zperf/src/zperf_shell.c
Comment (jukkar) [inline samples/net/zperf/src/zperf_shell.c]: iface is not needed, please remove it

File: samples/net/zperf/src/zperf_shell.c
Comment (jukkar) [inline samples/net/zperf/src/zperf_shell.c]: make this const so you do not need to do a cast below

File: samples/net/zperf/src/zperf_shell.c
Comment (jukkar) [inline samples/net/zperf/src/zperf_shell.c]: No need to initialize these two, and in fact the dst_in4_addr can be left out

File: samples/net/zperf/src/zperf_shell.c
Comment (jukkar) [inline samples/net/zperf/src/zperf_shell.c]: This should return const

File: samples/net/zperf/src/zperf_shell.c
Comment (jukkar) [inline samples/net/zperf/src/zperf_shell.c]: Same comments to this function as I had for IPv4.

File: samples/net/zperf/src/zperf_tcp_receiver.c
Comment (jukkar) [inline samples/net/zperf/src/zperf_tcp_receiver.c]: Why did you removed the empty string check?
Shouldn't this check look like
```
if (MY_IP4ADDR && MY_IPV4ADDR[0])
```

File: samples/net/zperf/src/zperf_tcp_receiver.c
Comment (jukkar) [inline samples/net/zperf/src/zperf_tcp_receiver.c]: same here

File: samples/net/zperf/src/zperf_udp_receiver.c
Comment (jukkar) [inline samples/net/zperf/src/zperf_udp_receiver.c]: and here

File: samples/net/zperf/src/zperf_udp_receiver.c
Comment (jukkar) [inline samples/net/zperf/src/zperf_udp_receiver.c]: and here

**Review by jukkar — CHANGES_REQUESTED** [review]

File: samples/net/zperf/src/zperf_tcp_receiver.c
Comment (bublover) [reply samples/net/zperf/src/zperf_tcp_receiver.c]: zephyr/samples/net/zperf/src/zperf_udp_receiver.c:329:30: warning: pointer of type ‘void *’ used in arithmetic [-Wpointer-arith]
if (MY_IP4ADDR && MY_IP4ADDR[0]) {
^
zephyr/samples/net/zperf/src/zperf_udp_receiver.c:329:30: error: void value not ignored as it ought to be
if (MY_IP4ADDR && MY_IP4ADDR[0]) {

File: samples/net/zperf/src/zperf_tcp_receiver.c
Comment (bublover) [reply samples/net/zperf/src/zperf_tcp_receiver.c]: Same reason becase of compile error if MY_IP6ADDR is NULL.

Comment (bublover) [reply samples/net/zperf/src/zperf_tcp_receiver.c]: > same here

Same reason becase of compile error if MY_IP6ADDR is NULL.

File: samples/net/zperf/src/zperf_udp_receiver.c
Comment (bublover) [reply samples/net/zperf/src/zperf_udp_receiver.c]: > and here

Same reason becase of compile error if MY_IP6ADDR is NULL.

File: samples/net/zperf/src/zperf_udp_receiver.c
Comment (bublover) [reply samples/net/zperf/src/zperf_udp_receiver.c]: > and here

Same reason becase of compile error if MY_IP6ADDR is NULL.

File: samples/net/zperf/src/zperf_tcp_receiver.c
Comment (bublover) [reply samples/net/zperf/src/zperf_tcp_receiver.c]: Hi @jukkar , could you pls give any idea?

Comment (jukkar) [reply samples/net/zperf/src/zperf_tcp_receiver.c]: As you said the MY_IP4ADDR can be set to NULL so the check cannot be done as I suggested. So we can do it like you proposed.

File: samples/net/zperf/src/zperf_shell.c
Comment (jukkar) [inline samples/net/zperf/src/zperf_shell.c]: We do not need dst_in4_addr variable, just use the net_ipv4_unspecified_address() directly in next line.
Also in4_addr is useless, just do
```
return net_if_ipv4_select_src_addr(NULL, net_ipv4_unspecified_address());
```
here.

File: samples/net/zperf/src/zperf_shell.c
Comment (jukkar) [inline samples/net/zperf/src/zperf_shell.c]: same comment here

**Review by bublover — COMMENTED** [review]
@jukkar , sure, removed unnecessary variables.
It seems nothing can be done further in this PR.
Thank you.

**Review by jukkar — APPROVED** [review]

**Pull request merged**
